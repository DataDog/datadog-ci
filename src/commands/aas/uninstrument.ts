import {WebSiteManagementClient} from '@azure/arm-appservice'
import {DefaultAzureCredential} from '@azure/identity'
import chalk from 'chalk'
import {Command} from 'clipanion'

import {renderError} from '../../helpers/renderer'

import {AAS_DD_SETTING_NAMES, AasCommand, formatError, isDotnet, parseEnvVars, SIDECAR_CONTAINER_NAME} from './common'
import {AasConfigOptions} from './interfaces'

export class UninstrumentCommand extends AasCommand {
  public static paths = [['aas', 'uninstrument']]
  public static usage = Command.Usage({
    category: 'Serverless',
    description: 'Remove Datadog instrumentation from an Azure App Service.',
  })

  public async execute(): Promise<0 | 1> {
    this.enableFips()
    const [appServicesToUninstrument, config, errors] = await this.ensureConfig()
    if (errors.length > 0) {
      for (const error of errors) {
        this.context.stdout.write(renderError(error))
      }

      return 1
    }

    const cred = new DefaultAzureCredential()
    if (!(await this.ensureAzureAuth(cred))) {
      return 1
    }
    this.context.stdout.write(`${this.dryRunPrefix}üê∂ Beginning uninstrumentation of Azure App Service(s)\n`)
    const results = await Promise.all(
      Object.entries(appServicesToUninstrument).map(([subscriptionId, resourceGroupToNames]) =>
        this.processSubscription(cred, subscriptionId, resourceGroupToNames, config)
      )
    )
    const success = results.every((result) => result)
    this.context.stdout.write(
      `${this.dryRunPrefix}üê∂ Uninstrumentation completed ${
        success ? 'successfully!' : 'with errors, see above for details.'
      }\n`
    )

    return success ? 0 : 1
  }

  public async processSubscription(
    cred: DefaultAzureCredential,
    subscriptionId: string,
    resourceGroupToNames: Record<string, string[]>,
    config: AasConfigOptions
  ): Promise<boolean> {
    const client = new WebSiteManagementClient(cred, subscriptionId, {apiVersion: '2024-11-01'})
    const results = await Promise.all(
      Object.entries(resourceGroupToNames).flatMap(([resourceGroup, aasNames]) =>
        aasNames.map((aasName) => this.processAas(client, config, resourceGroup, aasName))
      )
    )

    return results.every((result) => result)
  }

  /**
   * Process an Azure App Service for uninstrumentation.
   * @returns A promise that resolves to a boolean indicating success or failure.
   */
  public async processAas(
    client: WebSiteManagementClient,
    config: AasConfigOptions,
    resourceGroup: string,
    aasName: string
  ): Promise<boolean> {
    try {
      const site = await client.webApps.get(resourceGroup, aasName)
      if (!this.ensureLinux(site)) {
        return false
      }

      await this.uninstrumentSidecar(
        client,
        {...config, isDotnet: config.isDotnet || isDotnet(site)},
        resourceGroup,
        aasName
      )
    } catch (error) {
      this.context.stdout.write(renderError(`Failed to uninstrument ${chalk.bold(aasName)}: ${formatError(error)}`))

      return false
    }

    return true
  }

  public async uninstrumentSidecar(
    client: WebSiteManagementClient,
    config: AasConfigOptions,
    resourceGroup: string,
    aasName: string
  ) {
    this.context.stdout.write(
      `${this.dryRunPrefix}Removing sidecar container ${chalk.bold(SIDECAR_CONTAINER_NAME)} from ${chalk.bold(
        aasName
      )} (if it exists)\n`
    )
    if (!this.dryRun) {
      await client.webApps.deleteSiteContainer(resourceGroup, aasName, SIDECAR_CONTAINER_NAME)
    }
    const configuredSettings = new Set([...AAS_DD_SETTING_NAMES, ...Object.keys(parseEnvVars(config.envVars))])
    this.context.stdout.write(`${this.dryRunPrefix}Checking Application Settings on ${chalk.bold(aasName)}\n`)
    const currentEnvVars = (await client.webApps.listApplicationSettings(resourceGroup, aasName)).properties
    if (currentEnvVars !== undefined && Object.keys(currentEnvVars).some((key) => configuredSettings.has(key))) {
      this.context.stdout.write(`${this.dryRunPrefix}Updating Application Settings for ${chalk.bold(aasName)}\n`)
      if (!this.dryRun) {
        await client.webApps.updateApplicationSettings(resourceGroup, aasName, {
          properties: Object.fromEntries(
            Object.entries(currentEnvVars).filter(([key]) => !configuredSettings.has(key))
          ),
        })
      }
    } else {
      this.context.stdout.write(
        `${this.dryRunPrefix}No Application Settings changes needed for ${chalk.bold(aasName)}.\n`
      )
    }
  }
}
