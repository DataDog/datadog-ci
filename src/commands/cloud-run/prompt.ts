import inquirer from 'inquirer'

import {DATADOG_SITES} from '../../constants'

const checkboxPlusPrompt = require('inquirer-checkbox-plus-prompt')
inquirer.registerPrompt('checkbox-plus', checkboxPlusPrompt)

export const requestGCPProject = async (): Promise<string> => {
  const answer = await inquirer.prompt({
    message: 'Enter GCP Project ID:',
    name: 'project',
    type: 'input',
    validate: (value: string) => {
      if (!value || value.trim().length === 0) {
        return 'Project ID is required.'
      }

      return true
    },
  })

  return answer.project
}

export const requestGCPRegion = async (defaultRegion?: string): Promise<string> => {
  const answer = await inquirer.prompt({
    default: defaultRegion || 'us-central1',
    message: 'Enter GCP Region:',
    name: 'region',
    type: 'input',
    validate: (value: string) => {
      if (!value || value.trim().length === 0) {
        return 'Region is required.'
      }

      return true
    },
  })

  return answer.region
}

export const requestServiceName = async (): Promise<string> => {
  const answer = await inquirer.prompt({
    message: 'Enter Cloud Run service name:',
    name: 'serviceName',
    type: 'input',
    validate: (value: string) => {
      if (!value || value.trim().length === 0) {
        return 'Service name is required.'
      }

      return true
    },
  })

  return answer.serviceName
}

export const requestSite = async (): Promise<string> => {
  const answer = await inquirer.prompt({
    choices: DATADOG_SITES,
    message: 'Select a Datadog Site:',
    name: 'site',
    type: 'list',
  })

  return answer.site
}

export const requestConfirmation = async (message: string, defaultValue = true) => {
  const confirmationAnswer = await inquirer.prompt({
    message,
    name: 'confirmation',
    type: 'confirm',
    default: defaultValue,
  })

  return confirmationAnswer.confirmation !== false
}
