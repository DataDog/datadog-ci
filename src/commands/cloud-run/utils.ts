import type {IService, ServicesClient as IServicesClient} from './types'

import chalk from 'chalk'
import {GoogleAuth} from 'google-auth-library'
import {diff} from 'jest-diff'

import {withSpinner} from './renderer'

/**
 * Check if the user is authenticated with GCP.
 * @returns true if the user is authenticated, false otherwise
 */
export const checkAuthentication = async () => {
  const auth = new GoogleAuth()
  try {
    await auth.getApplicationDefault()

    return true
  } catch (_) {
    return false
  }
}

/**
 * Recursively sort object keys to ensure consistent ordering
 */
const sortObjectKeys = (obj: any): any => {
  if (!obj) {
    return obj
  }

  if (Array.isArray(obj)) {
    return obj.map(sortObjectKeys)
  }

  if (typeof obj === 'object') {
    const sorted: any = {}
    Object.keys(obj)
      .sort()
      .forEach((key) => {
        sorted[key] = sortObjectKeys(obj[key])
      })

    return sorted
  }

  return obj
}

/**
 * Obfuscate sensitive values in a line if it contains a key with "_KEY"
 */
const obfuscateSensitiveValues = (line: string): string => {
  // Match hex strings of 16, 32, or 64 characters (common API key/token lengths)
  return line
    .replace(/("[0-9a-fA-F]{16}"|"[0-9a-fA-F]{32}"|"[0-9a-fA-F]{64}")/g, '"***"')
    .replace(/('[0-9a-fA-F]{16}'|'[0-9a-fA-F]{32}'|'[0-9a-fA-F]{64}')/g, "'***'")
}

/**
 * Generate a git diff-style comparison between two configurations
 * TODO(@nhulston): update Lambda and AAS instrument to show this diff
 * @param original The original configuration object
 * @param updated The updated configuration object
 * @returns A formatted diff string with colors
 */
export const generateConfigDiff = (original: any, updated: any): string => {
  // Sort keys consistently before comparison
  const sortedOriginal = sortObjectKeys(original)
  const sortedUpdated = sortObjectKeys(updated)

  const originalJson = JSON.stringify(sortedOriginal, undefined, 2)
  const updatedJson = JSON.stringify(sortedUpdated, undefined, 2)

  const obfuscatedOriginal = originalJson.split('\n').map(obfuscateSensitiveValues).join('\n')
  const obfuscatedUpdated = updatedJson.split('\n').map(obfuscateSensitiveValues).join('\n')

  const configDiff = diff(obfuscatedOriginal, obfuscatedUpdated, {
    aColor: chalk.red,
    bColor: chalk.green,
    omitAnnotationLines: true,
  })
  if (!configDiff || configDiff.includes('no visual difference')) {
    return chalk.gray('No changes detected.')
  }

  return configDiff
}

export const fetchServiceConfigs = async (
  client: IServicesClient,
  project: string,
  region: string,
  services: string[]
) => {
  const existingServiceConfigs: IService[] = []
  for (const serviceName of services) {
    const servicePath = client.servicePath(project, region, serviceName)

    const existingService = await withSpinner(
      `Fetching configuration for ${chalk.bold(serviceName)}...`,
      async () => {
        try {
          const [serv] = await client.getService({name: servicePath})

          return serv
        } catch (error) {
          throw new Error(
            `Service ${serviceName} not found in project ${project}, region ${region}.\n\nNo services were instrumented.\n`
          )
        }
      },
      `Fetched service configuration for ${chalk.bold(serviceName)}`
    )
    existingServiceConfigs.push(existingService)
  }

  return existingServiceConfigs
}
