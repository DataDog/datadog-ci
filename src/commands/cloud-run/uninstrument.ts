import type {IContainer, IService, IVolume, ServicesClient as IServicesClient} from './types'

import chalk from 'chalk'
import {Command, Option} from 'clipanion'

import {FIPS_ENV_VAR, FIPS_IGNORE_ERROR_ENV_VAR} from '../../constants'
import {toBoolean} from '../../helpers/env'
import {enableFips} from '../../helpers/fips'
import {renderError, renderSoftWarning} from '../../helpers/renderer'

import {DEFAULT_SIDECAR_NAME, DEFAULT_VOLUME_NAME} from './constants'
import {requestGCPProject, requestGCPRegion, requestServiceName, requestConfirmation} from './prompt'
import {dryRunPrefix, renderAuthenticationInstructions, withSpinner} from './renderer'
import {checkAuthentication, fetchServiceConfigs, generateConfigDiff} from './utils'

// XXX temporary workaround for @google-cloud/run ESM/CJS module issues
const {ServicesClient} = require('@google-cloud/run')

export class UninstrumentCommand extends Command {
  public static paths = [['cloud-run', 'uninstrument']]

  public static usage = Command.Usage({
    category: 'Serverless',
    description: 'Revert Datadog instrumentation in a Cloud Run app.',
  })

  private dryRun = Option.Boolean('-d,--dry,--dry-run', false)
  private project = Option.String('-p,--project', {
    description: 'GCP project ID',
  })
  private services = Option.Array('-s,--service,--services', [], {
    description: 'Cloud Run service(s) to instrument',
  })
  private interactive = Option.Boolean('-i,--interactive', false, {
    description: 'Prompt for flags one at a time',
  })
  private region = Option.String('-r,--region', {
    description: 'GCP region your service(s) are deployed in',
  })
  private sidecarName = Option.String('--sidecar-name', DEFAULT_SIDECAR_NAME, {
    description: `The name of the sidecar container to remove. Specify if you have a different sidecar name. Defaults to '${DEFAULT_SIDECAR_NAME}'`,
  })
  private sharedVolumeName = Option.String('--shared-volume-name', DEFAULT_VOLUME_NAME, {
    description: `The name of the shared volume to remove. Specify if you have a different shared volume name. Defaults to '${DEFAULT_VOLUME_NAME}'`,
  })
  private fips = Option.Boolean('--fips', false)
  private fipsIgnoreError = Option.Boolean('--fips-ignore-error', false)

  private fipsConfig = {
    fips: toBoolean(process.env[FIPS_ENV_VAR]) ?? false,
    fipsIgnoreError: toBoolean(process.env[FIPS_IGNORE_ERROR_ENV_VAR]) ?? false,
  }

  public async execute(): Promise<0 | 1> {
    enableFips(this.fips || this.fipsConfig.fips, this.fipsIgnoreError || this.fipsConfig.fipsIgnoreError)

    this.context.stdout.write(
      `\n${dryRunPrefix(this.dryRun)}üê∂ ${chalk.bold('Uninstrumenting Cloud Run service(s)')}\n\n`
    )

    if (this.interactive) {
      if (!this.project) {
        this.project = await requestGCPProject()
      }

      if (!this.region) {
        this.region = await requestGCPRegion()
      }

      if (this.services.length === 0) {
        const serviceName = await requestServiceName()
        this.services = [serviceName]
      }
    }

    // Validate required variables
    if (!this.project) {
      this.context.stdout.write(chalk.yellow('Invalid or missing project. Please use the --project flag.\n'))
    }
    if (this.services.length === 0) {
      this.context.stdout.write(chalk.yellow('Invalid or missing service(s). Please use the --service flag.\n'))
    }
    if (!this.region) {
      this.context.stdout.write(chalk.yellow('Invalid or missing region. Please use the --region flag.\n'))
    }

    if (!this.project || !this.services || !this.services.length || !this.region) {
      return 1
    }
    this.context.stdout.write(chalk.green('‚úî Required flags verified\n'))

    // Verify GCP credentials
    this.context.stdout.write(chalk.bold('\nüîë Verifying GCP credentials...\n'))
    const authenticated = await checkAuthentication()
    if (!authenticated) {
      this.context.stderr.write(renderAuthenticationInstructions())

      return 1
    }
    this.context.stdout.write(chalk.green('‚úî GCP credentials verified!\n\n'))

    // Instrument services with sidecar
    try {
      await this.uninstrumentSidecar(this.project, this.services, this.region)
    } catch (error) {
      this.context.stderr.write(dryRunPrefix(this.dryRun) + renderError(`Uninstrumentation failed: ${error}\n`))

      return 1
    }

    if (!this.dryRun) {
      this.context.stdout.write('\n‚úÖ Cloud Run uninstrumentation completed successfully!\n')
    }

    return 0
  }

  public async uninstrumentSidecar(project: string, services: string[], region: string) {
    const client: IServicesClient = new ServicesClient()

    this.context.stdout.write(
      chalk.bold(`\n${dryRunPrefix(this.dryRun)}‚¨áÔ∏è Fetching existing service configurations from Cloud Run...\n`)
    )
    const existingServiceConfigs = await fetchServiceConfigs(client, project, region, services)

    this.context.stdout.write(
      chalk.bold(`\n${dryRunPrefix(this.dryRun)}üöÄ Uninstrumenting Cloud Run services with sidecar...\n`)
    )
    for (let i = 0; i < existingServiceConfigs.length; i++) {
      const serviceConfig = existingServiceConfigs[i]
      const serviceName = services[i]
      try {
        await this.uninstrumentService(client, serviceConfig, serviceName)
      } catch (error) {
        this.context.stderr.write(
          dryRunPrefix(this.dryRun) + renderError(`Failed to instrument service ${serviceName}: ${error}\n`)
        )
        throw error
      }
    }
  }

  public async uninstrumentService(client: IServicesClient, existingService: IService, serviceName: string) {
    const updatedService = this.createUninstrumentedServiceConfig(existingService)
    this.context.stdout.write(generateConfigDiff(existingService, updatedService))
    if (this.dryRun) {
      this.context.stdout.write(
        `\n\n${dryRunPrefix(this.dryRun)}Would have updated service ${chalk.bold(
          serviceName
        )} with the above changes.\n`
      )

      return
    } else if (this.interactive) {
      const confirmed = await requestConfirmation('\nDo you want to apply the changes?')
      if (!confirmed) {
        throw new Error('Uninstrumentation cancelled by user.')
      }
    }

    await withSpinner(
      `Uninstrumenting service ${chalk.bold(serviceName)}...`,
      async () => {
        const [operation] = await client.updateService({
          service: updatedService,
        })
        await operation.promise()
      },
      `Uninstrumented service ${chalk.bold(serviceName)}`
    )
  }

  public createUninstrumentedServiceConfig(service: IService): IService {
    const template = service.template || {}
    const containers: IContainer[] = template.containers || []
    const volumes: IVolume[] = template.volumes || []

    let updatedContainers = containers.filter((c) => c.name !== this.sidecarName)
    const updatedVolumes = volumes.filter((v) => v.name !== this.sharedVolumeName)

    if (updatedContainers.length === containers.length) {
      this.context.stdout.write(
        renderSoftWarning(`Sidecar container '${this.sidecarName}' not found, so no container was removed. Specify the container name with --sidecar-name.
`)
      )
    }

    if (updatedVolumes.length === volumes.length) {
      this.context.stdout.write(
        renderSoftWarning(`Shared volume '${this.sharedVolumeName}' not found, so no shared volume was removed. Specify the shared volume name with --shared-volume-name.
`)
      )
    }

    updatedContainers = updatedContainers.map((c) => this.updateAppContainer(c))

    return {
      ...service,
      template: {
        ...template,
        containers: updatedContainers,
        volumes: updatedVolumes,
        // Let GCR generate the next revision name
        revision: undefined,
      },
    }
  }

  // Remove volume mount and add required env vars
  private updateAppContainer(appContainer: IContainer) {
    const existingVolumeMounts = appContainer.volumeMounts || []
    const updatedVolumeMounts = existingVolumeMounts.filter((v) => v.name !== this.sharedVolumeName)

    const existingEnvVars = appContainer.env || []
    // Remove env vars beginning with DD_
    const updatedEnvVars = existingEnvVars.filter((v) => !v.name.startsWith('DD_'))

    return {
      ...appContainer,
      volumeMounts: updatedVolumeMounts,
      env: updatedEnvVars,
    }
  }
}
