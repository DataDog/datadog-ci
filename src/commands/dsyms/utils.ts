import {exec} from 'child_process'
import {promises} from 'fs'
import {tmpdir} from 'os'
import path from 'path'
import {promisify} from 'util'

import rimraf from 'rimraf'

import {buildPath} from '../../helpers/utils'

export const isZipFile = async (filepath: string) => {
  try {
    const stats = await promises.stat(filepath)

    return stats.size !== 0 && path.extname(filepath) === '.zip'
  } catch (error) {
    // Log to console "file exists yet empty" ?
    return false
  }
}

export const createUniqueTmpDirectory = async (): Promise<string> => {
  const uniqueValue = Math.random() * Number.MAX_SAFE_INTEGER
  const directoryPath = buildPath(tmpdir(), uniqueValue.toString())
  await promises.mkdir(directoryPath, {recursive: true})

  return directoryPath
}

export const deleteDirectory = async (directoryPath: string): Promise<void> =>
  new Promise((resolve, reject) => rimraf(directoryPath, () => resolve()))

export const zipDirectoryToArchive = async (directoryPath: string, archivePath: string) => {
  const cwd = path.dirname(directoryPath)
  const directoryName = path.basename(directoryPath)
  await execute(`zip -r '${archivePath}' '${directoryName}'`, cwd)
}

export const unzipArchiveToDirectory = async (archivePath: string, directoryPath: string) => {
  await promises.mkdir(directoryPath, {recursive: true})
  await execute(`unzip -o '${archivePath}' -d '${directoryPath}'`)
}

export const executeDwarfdump = async (dSYMPath: string): Promise<{stderr: string; stdout: string}> =>
  execute(`dwarfdump --uuid '${dSYMPath}'`)

export const executeLipo = async (
  objectPath: string,
  arch: string,
  newObjectPath: string
): Promise<{stderr: string; stdout: string}> => execute(`lipo '${objectPath}' -thin ${arch} -output '${newObjectPath}'`)

const execProc = promisify(exec)
const execute = (cmd: string, cwd?: string): Promise<{stderr: string; stdout: string}> =>
  execProc(cmd, {
    cwd,
    maxBuffer: 5 * 1024 * 5000,
  })

export const getBaseIntakeUrl = (datadogSite?: string) => {
  if (process.env.DATADOG_DSYM_INTAKE_URL) {
    return process.env.DATADOG_DSYM_INTAKE_URL
  } else if (datadogSite) {
    return 'https://sourcemap-intake.' + datadogSite
  }

  return 'https://sourcemap-intake.datadoghq.com'
}

export const pluralize = (nb: number, singular: string, plural: string) => {
  if (nb >= 2) {
    return `${nb} ${plural}`
  }

  return `${nb} ${singular}`
}
