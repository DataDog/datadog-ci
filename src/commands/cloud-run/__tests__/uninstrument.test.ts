// XXX temporary workaround for @google-cloud/run ESM/CJS module issues
import type {IContainer} from '../types'

import {
  API_KEY_ENV_VAR,
  ENVIRONMENT_ENV_VAR,
  DD_TAGS_ENV_VAR,
  DD_TRACE_ENABLED_ENV_VAR,
  SERVICE_ENV_VAR,
} from '../../../constants'
import {makeRunCLI} from '../../../helpers/__tests__/testing-tools'

import * as cloudRunPromptModule from '../prompt'
import {UninstrumentCommand} from '../uninstrument'
import * as utils from '../utils'

jest.mock('../utils', () => ({
  ...jest.requireActual('../utils'),
  checkAuthentication: jest.fn(),
}))

const mockServicesClient = {
  servicePath: jest.fn(),
  getService: jest.fn(),
  updateService: jest.fn(),
}

jest.mock('@google-cloud/run', () => ({
  ServicesClient: jest.fn(() => mockServicesClient),
}))

describe('UninstrumentCommand', () => {
  const runCLI = makeRunCLI(UninstrumentCommand, ['cloud-run', 'uninstrument'])

  beforeEach(() => {
    jest.clearAllMocks()
    ;(utils.checkAuthentication as jest.Mock).mockResolvedValue(true)
    mockServicesClient.servicePath.mockImplementation(
      (project: string, region: string, service: string) =>
        `projects/${project}/locations/${region}/services/${service}`
    )
  })

  describe('validates required variables', () => {
    test('should fail if required flags are missing', async () => {
      const cases = [
        {args: ['--services', 'test-service', '--region', 'us-central1'], expected: 'missing project'},
        {args: ['--project', 'test-project', '--region', 'us-central1'], expected: 'missing service(s)'},
        {args: ['--project', 'test-project', '--services', 'test-service'], expected: 'missing region'},
      ]

      for (const {args, expected} of cases) {
        const {code, context} = await runCLI(args)
        expect(code).toBe(1)
        expect(context.stdout.toString()).toContain(expected)
      }
    })
  })

  describe('main uninstrument command flow', () => {
    test('should fail if GCP credentials are invalid', async () => {
      ;(utils.checkAuthentication as jest.Mock).mockResolvedValue(false)
      const {code, context} = await runCLI([
        '--project',
        'test-project',
        '--services',
        'test-service',
        '--region',
        'us-central1',
      ])
      expect(code).toBe(1)
      expect(context.stderr.toString()).toContain('Unable to authenticate with GCP')
    })

    test('should succeed with valid parameters', async () => {
      const mockUninstrumentSidecar = jest.fn().mockResolvedValue(undefined)
      jest
        .spyOn(UninstrumentCommand.prototype as any, 'uninstrumentSidecar')
        .mockImplementation(mockUninstrumentSidecar)

      const {code} = await runCLI([
        '--project',
        'test-project',
        '--services',
        'test-service',
        '--region',
        'us-central1',
      ])
      expect(code).toBe(0)
      expect(mockUninstrumentSidecar).toHaveBeenCalledWith('test-project', ['test-service'], 'us-central1')
    })
  })

  describe('snapshot tests', () => {
    const mockService = {
      name: 'projects/test-project/locations/us-central1/services/test-service',
      template: {
        containers: [
          {
            name: 'main-app',
            image: 'gcr.io/test-project/test-app:latest',
            env: [
              {name: 'NODE_ENV', value: 'production'},
              {name: DD_TRACE_ENABLED_ENV_VAR, value: 'true'},
              {name: SERVICE_ENV_VAR, value: 'test-service'},
            ],
            volumeMounts: [{name: 'shared-volume', mountPath: '/shared-volume'}],
          },
          {
            name: 'datadog-sidecar',
            image: 'gcr.io/datadoghq/serverless-init:latest',
            env: [{name: API_KEY_ENV_VAR, value: 'test-api-key'}],
            volumeMounts: [{name: 'shared-volume', mountPath: '/shared-volume'}],
          },
        ],
        volumes: [{name: 'shared-volume', emptyDir: {}}],
        revision: 'test-service-v1',
      },
    }

    beforeEach(() => {
      mockServicesClient.getService.mockResolvedValue([mockService])
      mockServicesClient.updateService.mockResolvedValue([{promise: jest.fn().mockResolvedValue([])}])
      jest.restoreAllMocks()
      ;(utils.checkAuthentication as jest.Mock).mockResolvedValue(true)
    })

    test('prints dry run data', async () => {
      const {code, context} = await runCLI([
        '--project',
        'test-project',
        '--services',
        'test-service',
        '--region',
        'us-central1',
        '--dry-run',
      ])
      expect(code).toBe(0)
      expect(context.stdout.toString()).toMatchSnapshot()
    })

    test('interactive mode', async () => {
      jest.spyOn(cloudRunPromptModule, 'requestGCPProject').mockResolvedValue('interactive-project')
      jest.spyOn(cloudRunPromptModule, 'requestGCPRegion').mockResolvedValue('us-west1')
      jest.spyOn(cloudRunPromptModule, 'requestServiceName').mockResolvedValue('interactive-service')
      jest.spyOn(cloudRunPromptModule, 'requestConfirmation').mockResolvedValue(true)

      const {code, context} = await runCLI(['--interactive'])
      expect(code).toBe(0)
      expect(context.stdout.toString()).toMatchSnapshot()
    })
  })

  describe('createUninstrumentedServiceConfig', () => {
    let command: UninstrumentCommand

    beforeEach(() => {
      command = new UninstrumentCommand()
      ;(command as any).sidecarName = 'datadog-sidecar'
      ;(command as any).sharedVolumeName = 'shared-volume'
      ;(command as any).context = {
        stdout: {write: jest.fn()},
        stderr: {write: jest.fn()},
      }
    })

    test('removes sidecar container, shared volume, and DD_ env vars', () => {
      const service = {
        template: {
          containers: [
            {
              name: 'main',
              env: [
                {name: 'NODE_ENV', value: 'production'},
                {name: DD_TRACE_ENABLED_ENV_VAR, value: 'true'},
                {name: ENVIRONMENT_ENV_VAR, value: 'staging'},
                {name: DD_TAGS_ENV_VAR, value: 'team:backend'},
                {name: 'CUSTOM_VAR', value: 'keep-me'},
              ],
              volumeMounts: [{name: 'shared-volume', mountPath: '/shared-volume'}],
            },
            {
              name: 'datadog-sidecar',
              env: [{name: API_KEY_ENV_VAR, value: 'test-key'}],
              volumeMounts: [{name: 'shared-volume', mountPath: '/shared-volume'}],
            },
          ],
          volumes: [{name: 'shared-volume', emptyDir: {}}],
        },
      }

      const result = command.createUninstrumentedServiceConfig(service)

      // Should remove sidecar container
      expect(result.template?.containers).toHaveLength(1)
      expect(result.template?.containers?.map((c: IContainer) => c.name)).toEqual(['main'])

      // Should remove shared volume
      expect(result.template?.volumes).toHaveLength(0)

      // Should remove DD_ env vars and shared volume mount from main container
      const main = result.template?.containers?.find((c: IContainer) => c.name === 'main')
      expect(main?.volumeMounts).toHaveLength(0)
      expect(main?.env).toEqual([
        {name: 'NODE_ENV', value: 'production'},
        {name: 'CUSTOM_VAR', value: 'keep-me'},
      ])
    })

    test('handles service with no sidecar or shared volume gracefully', () => {
      const service = {
        template: {
          containers: [{name: 'main', env: [{name: 'NODE_ENV', value: 'production'}], volumeMounts: []}],
          volumes: [],
        },
      }

      const result = command.createUninstrumentedServiceConfig(service)
      expect(result.template?.containers).toHaveLength(1)
      expect(result.template?.volumes).toHaveLength(0)
    })

    test('uses custom sidecar and volume names', () => {
      ;(command as any).sidecarName = 'custom-sidecar'
      ;(command as any).sharedVolumeName = 'custom-volume'

      const service = {
        template: {
          containers: [
            {name: 'main', env: [], volumeMounts: [{name: 'custom-volume', mountPath: '/custom/path'}]},
            {name: 'custom-sidecar', env: [], volumeMounts: []},
          ],
          volumes: [{name: 'custom-volume', emptyDir: {}}],
        },
      }

      const result = command.createUninstrumentedServiceConfig(service)
      expect(result.template?.containers).toHaveLength(1)
      expect(result.template?.containers?.map((c: IContainer) => c.name)).toEqual(['main'])
      expect(result.template?.volumes).toHaveLength(0)
    })
  })

  describe('updateAppContainer', () => {
    let command: UninstrumentCommand

    beforeEach(() => {
      command = new UninstrumentCommand()
      ;(command as any).sharedVolumeName = 'shared-volume'
      ;(command as any).context = {
        stdout: {write: jest.fn()},
        stderr: {write: jest.fn()},
      }
    })

    test('removes shared volume mount and DD_ environment variables', () => {
      const appContainer = {
        name: 'main',
        env: [
          {name: 'NODE_ENV', value: 'production'},
          {name: DD_TRACE_ENABLED_ENV_VAR, value: 'true'},
          {name: DD_TAGS_ENV_VAR, value: 'team:backend'},
          {name: 'CUSTOM_VAR', value: 'keep-me'},
        ],
        volumeMounts: [
          {name: 'shared-volume', mountPath: '/shared-volume'},
          {name: 'other-volume', mountPath: '/other'},
        ],
      }

      const result = (command as any).updateAppContainer(appContainer)

      expect(result.volumeMounts).toEqual([{name: 'other-volume', mountPath: '/other'}])
      expect(result.env).toEqual([
        {name: 'NODE_ENV', value: 'production'},
        {name: 'CUSTOM_VAR', value: 'keep-me'},
      ])
    })

    test('handles container with undefined env and volumeMounts', () => {
      const appContainer = {name: 'main'}
      const result = (command as any).updateAppContainer(appContainer)

      expect(result.volumeMounts).toEqual([])
      expect(result.env).toEqual([])
    })
  })
})
