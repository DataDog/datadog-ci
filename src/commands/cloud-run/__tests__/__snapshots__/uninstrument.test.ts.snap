// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UninstrumentCommand snapshot tests interactive mode 1`] = `
"
üê∂ Uninstrumenting Cloud Run service(s)

‚úî Required flags verified

üîë Verifying GCP credentials...
‚úî GCP credentials verified!


‚¨áÔ∏è Fetching existing service configurations from Cloud Run...

üöÄ Uninstrumenting Cloud Run services with sidecar...
  {
    "name": "projects/test-project/locations/us-central1/services/test-service",
    "template": {
      "containers": [
        {
          "env": [
            {
              "name": "NODE_ENV",
              "value": "production"
-           },
-           {
-             "name": "DD_TRACE_ENABLED",
-             "value": "true"
-           },
-           {
-             "name": "DD_SERVICE",
-             "value": "test-service"
            }
          ],
          "image": "gcr.io/test-project/test-app:latest",
          "name": "main-app",
-         "volumeMounts": [
-           {
-             "mountPath": "/shared-volume",
-             "name": "shared-volume"
-           }
-         ]
-       },
-       {
-         "env": [
-           {
-             "name": "DD_API_KEY",
-             "value": "test-api-key"
-           }
-         ],
-         "image": "gcr.io/datadoghq/serverless-init:latest",
-         "name": "datadog-sidecar",
-         "volumeMounts": [
-           {
-             "mountPath": "/shared-volume",
-             "name": "shared-volume"
-           }
-         ]
+         "volumeMounts": []
        }
      ],
-     "revision": "test-service-v1",
-     "volumes": [
-       {
-         "emptyDir": {},
-         "name": "shared-volume"
-       }
-     ]
+     "volumes": []
    }
  }
‚úÖ Cloud Run uninstrumentation completed successfully!
"
`;

exports[`UninstrumentCommand snapshot tests prints dry run data 1`] = `
"
[Dry Run] üê∂ Uninstrumenting Cloud Run service(s)

‚úî Required flags verified

üîë Verifying GCP credentials...
‚úî GCP credentials verified!


[Dry Run] ‚¨áÔ∏è Fetching existing service configurations from Cloud Run...

[Dry Run] üöÄ Uninstrumenting Cloud Run services with sidecar...
  {
    "name": "projects/test-project/locations/us-central1/services/test-service",
    "template": {
      "containers": [
        {
          "env": [
            {
              "name": "NODE_ENV",
              "value": "production"
-           },
-           {
-             "name": "DD_TRACE_ENABLED",
-             "value": "true"
-           },
-           {
-             "name": "DD_SERVICE",
-             "value": "test-service"
            }
          ],
          "image": "gcr.io/test-project/test-app:latest",
          "name": "main-app",
-         "volumeMounts": [
-           {
-             "mountPath": "/shared-volume",
-             "name": "shared-volume"
-           }
-         ]
-       },
-       {
-         "env": [
-           {
-             "name": "DD_API_KEY",
-             "value": "test-api-key"
-           }
-         ],
-         "image": "gcr.io/datadoghq/serverless-init:latest",
-         "name": "datadog-sidecar",
-         "volumeMounts": [
-           {
-             "mountPath": "/shared-volume",
-             "name": "shared-volume"
-           }
-         ]
+         "volumeMounts": []
        }
      ],
-     "revision": "test-service-v1",
-     "volumes": [
-       {
-         "emptyDir": {},
-         "name": "shared-volume"
-       }
-     ]
+     "volumes": []
    }
  }

[Dry Run] Would have updated service test-service with the above changes.
"
`;
